src/notifications/notificationmanager.cpp

```diff
@@ -70,28 +71,92 @@ const int DefaultNotificationPriority = 50;
 const int CommitDelay = 10 * 1000;
 const int PublicationDelay = 1000;
 
-QString getProcessName(uint pid)
+bool processIsPrivileged(int pid)
 {
-    if (pid > 1) {
-        const QString procFilename(QString::fromLatin1("/proc/%1/cmdline").arg(QString::number(pid)));
-        QFile procFile(procFilename);
-        if (procFile.open(QIODevice::ReadOnly)) {
-            const QByteArray cmdLine = procFile.readAll();
-            QString processNameWithPath = QFile::decodeName(cmdLine.left(cmdLine.indexOf('\0')));
-            return QFileInfo(processNameWithPath).fileName();
+    bool isPrivileged = false;
+    if (pid == getpid()) {
+        // Internal operations are considered privileged
+        isPrivileged = true;
+    } else if (pid > 0) {
+        QFileInfo info(QString("/proc/%1").arg(pid));
+        isPrivileged = info.group() == QLatin1String("privileged") || info.owner() == QLatin1String("root");
+    }
+    NOTIFICATIONS_DEBUG("pid" << pid << "-> isPrivileged" << isPrivileged);
+    return isPrivileged;
+}
+
+QString getProcessCmdline(int pid)
+{
+    QString cmdline;
+    if (pid > 0) {
+        const QString path(QString::fromLatin1("/proc/%1/cmdline").arg(QString::number(pid)));
+        QFile file(path);
+        if (file.open(QIODevice::ReadOnly)) {
+            const QByteArray data = file.readAll();
+            cmdline = QString::fromUtf8(data.left(data.indexOf('\0')));
         }
     }
-    return QString();
+    NOTIFICATIONS_DEBUG("pid" << pid << "-> cmdline" << cmdline);
+    return cmdline;
 }
 
-QPair<QString, QString> processProperties(uint pid)
+QString getProcessComm(int pid)
+{
+    QString comm;
+    if (pid > 0) {
+        const QString path(QString::fromLatin1("/proc/%1/comm").arg(QString::number(pid)));
+        QFile file(path);
+        if (file.open(QIODevice::ReadOnly)) {
+            const QByteArray data = file.readAll();
+            comm = QFile::decodeName(data.left(data.indexOf('\n')));
+        }
+    }
+    NOTIFICATIONS_DEBUG("pid" << pid << "-> comm" << comm);
+    return comm;
+}
+
+QString getProcessExe(int pid)
+{
+    QString exe;
+    if (pid > 0) {
+        const QString path(QString::fromLatin1("/proc/%1/exe").arg(QString::number(pid)));
+        exe = QFile::symLinkTarget(path);
+    }
+    NOTIFICATIONS_DEBUG("pid" << pid << "-> exe" << exe);
+    return exe;
+}
+
+bool processIsDBusProxy(int pid)
+{
+    bool isDBusProxy = false;
+    if (pid > 0) {
+        static const char proxyPath[] = "/usr/bin/xdg-dbus-proxy";
+        static const char proxyName[] = "xdg-dbus-proxy";
+        const QString exe(getProcessExe(pid));
+        if (!exe.isEmpty())
+            isDBusProxy = exe == proxyPath;
+        else
+            isDBusProxy = getProcessComm(pid) == proxyName || getProcessCmdline(pid) == proxyPath;
```

Comment:
Perhaps a comment for future code archaeologists about this not being trustworthy security-wise would be in order so that this code will not get used for such purposes by accident.

Reply:
Good point. Added an explanatory comment.